diff --git a/utils/tinymix.c b/utils/tinymix.c
index f0e1d95..3d4fb63 100644
--- a/utils/tinymix.c
+++ b/utils/tinymix.c
@@ -219,7 +219,7 @@ static void list_controls(struct mixer *mixer, int print_all)
         name = mixer_ctl_get_name(ctl);
         type = mixer_ctl_get_type_string(ctl);
         num_values = mixer_ctl_get_num_values(ctl);
-        device = mixer_ctl_get_dev_num(ctl);
+        device = mixer_ctl_get_device(ctl);
         printf("%u\t%s\t%u\t%-40s\t%u", i, type, num_values, name, device);
         if (print_all)
             print_control_values(ctl);
@@ -578,4 +578,3 @@ static int set_values(struct mixer *mixer, const char *control,
 
     return 0;
 }
-
