diff --git a/configure.ac b/configure.ac
index e05f95b..8af2196 100644
--- a/configure.ac
+++ b/configure.ac
@@ -12,6 +12,9 @@ This file is under the same license as Axel.
 
 AC_PREREQ([2.69])
 PKG_PREREQ([0.29])
+m4_ifdef([AX_IS_RELEASE],, [
+    m4_fatal([autoconf-archive is required])
+])
 
 AXEL_PKG([axel], [https://github.com/axel-download-accelerator/axel/issues])
 AXEL_VER_READ([VERSION])
@@ -22,11 +25,24 @@ AC_CONFIG_SRCDIR([src/conf.h])
 AC_CONFIG_HEADERS([config.h])
 AC_CONFIG_MACRO_DIR([m4])
 AC_CONFIG_LIBOBJ_DIR([lib])
+
+dnl Validate arguments; fail early
+AC_ARG_WITH([ssl], [
+AS_HELP_STRING(
+        [--with-ssl=@<:@openssl/wolfssl@:>@],
+        [Which TLS/SSL implementation to use, default: openssl])
+AS_HELP_STRING(
+        [--without-ssl], [disable TLS support])],
+    [AS_CASE(["$withval"],
+        [openssl|wolfssl|no],,
+        [AC_MSG_ERROR([Invalid argument: --with-ssl=$withval])])],
+    [with_ssl=openssl])
+
 AM_INIT_AUTOMAKE([subdir-objects])
 PKG_PROG_PKG_CONFIG
 AS_IF([test -z "$PKG_CONFIG"],
-  [AC_MSG_WARN([pkg-config is needed for auto-configuration of dependencies])
-   PKG_CONFIG=false])
+  [AC_MSG_WARN([pkg-config is needed for auto-configuration of dependencies])])
+export PKG_CONFIG_PATH
 
 AM_SILENT_RULES([yes])
 
@@ -175,24 +191,40 @@ AC_REPLACE_FUNCS([ \
 AXEL_CHECK_MACRO([O_NONBLOCK], [fcntl.h])
 
 # Optional (but included-by-default) ssl support
-AC_ARG_WITH([ssl],
-    AC_HELP_STRING([--without-ssl],[disable TLS support]), [], [with_ssl=yes])
-
 AM_CONDITIONAL([WITH_SSL], [test "x$with_ssl" != xno])
+AC_ARG_VAR([SSL_PREFIX], [installation prefix of the TLS/SSL library])
 AS_IF([test "x$with_ssl" != xno], [
-    AS_IF([test "x$with_ssl" = xyes], [], [
-        PKG_CONFIG_PATH="$with_ssl/lib/pkgconfig$PATH_SEPARATOR$PKG_CONFIG_PATH"
-        export PKG_CONFIG_PATH
+    save_PKG_CONFIG_PATH="$PKG_CONFIG_PATH"
+    AS_IF([test ${SSL_PREFIX+y}], [
+	PKG_CONFIG_PATH="$SSL_PREFIX/lib/pkgconfig$PATH_SEPARATOR$PKG_CONFIG_PATH"
     ])
-    PKG_CHECK_MODULES([SSL], [openssl])
+    PKG_CHECK_MODULES([SSL], [$with_ssl])
     AC_DEFINE([HAVE_SSL], [1], [SSL])
-    save_LIBS="$LIBS"
-    LIBS="$LIBS $SSL_LIBS"
-    AC_REPLACE_FUNCS([ \
-        ASN1_STRING_get0_data \
-    ])
-    LIBS="$save_LIBS"
-], AC_MSG_NOTICE([SSL support disabled]))
+    AS_CASE(["$with_ssl"],
+	[openssl], [
+	    save_LIBS="$LIBS"
+	    LIBS="$LIBS $SSL_LIBS"
+	    AC_REPLACE_FUNCS([ \
+		ASN1_STRING_get0_data \
+	    ])
+	    LIBS="$save_LIBS"
+	],
+	[wolfssl], [
+	    AC_CHECK_DECLS([
+		X509_NAME_get_entry,
+		X509_get_ext_d2i,
+	        X509_V_OK
+	    ],, [
+	        AC_MSG_ERROR([wolfSSL with extra OpenSSL API compatibility required])
+	    ], [
+	        #include <wolfssl/options.h>
+		#include <wolfssl/wolfcrypt/settings.h>
+		#include <wolfssl/openssl/ssl.h>
+	    ])
+	    AC_DEFINE([HAVE_WOLFSSL], [1], [wolfSSL])
+	])
+    PKG_CONFIG_PATH="$save_PKG_CONFIG_PATH"
+], AC_MSG_NOTICE([TLS/SSL support disabled]))
 
 # Add Gettext
 AM_GNU_GETTEXT([external])
diff --git a/lgtm.yml b/lgtm.yml
new file mode 100644
index 0000000..714bd12
--- /dev/null
+++ b/lgtm.yml
@@ -0,0 +1,15 @@
+extraction:
+  cpp:
+    prepare:
+      packages:
+        - autoconf-archive
+        - autopoint
+        - clang
+        - libssl-dev
+        - linux-libc-dev
+        - txt2man
+        - gettext
+    configure:
+      command:
+      - autoreconf -ivf
+      - ./configure
diff --git a/lib/ASN1_STRING_get0_data.c b/lib/ASN1_STRING_get0_data.c
index 4ed0735..9f6f833 100644
--- a/lib/ASN1_STRING_get0_data.c
+++ b/lib/ASN1_STRING_get0_data.c
@@ -4,8 +4,18 @@
  * SPDX-License-Identifier: GPL-2.0-or-later
  */
 
+#include "config.h"
+
+#ifdef HAVE_WOLFSSL
+#include <wolfssl/options.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/openssl/ssl.h>
+#include <wolfssl/openssl/x509v3.h>
+#else
 #include <openssl/ssl.h>
 #include <openssl/x509v3.h>
+#endif
+
 #include "compat-ssl.h"
 
 const unsigned char *
diff --git a/po/tr.po b/po/tr.po
index 834a309..27b86be 100644
--- a/po/tr.po
+++ b/po/tr.po
@@ -52,7 +52,7 @@ msgstr "Dosya boyutu: müsait değil"
 #: src/axel.c:269
 #, c-format
 msgid "Opening output file %s"
-msgstr "%s çıktı dosyası açılıyor"
+msgstr "%s çıktı dosyası oluşturuluyor"
 
 #: src/axel.c:276
 msgid "Server unsupported, starting from scratch with one connection."
diff --git a/src/Makefile.am b/src/Makefile.am
index 5ac8e24..1826ecf 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -15,8 +15,11 @@ axel_SOURCES = \
 	src/conn.h \
 	src/ftp.c \
 	src/ftp.h \
+	src/hash.c \
+	src/hash.h \
 	src/http.c \
 	src/http.h \
+	src/random.c \
 	src/search.c \
 	src/search.h \
 	src/ssl.h \
@@ -24,18 +27,20 @@ axel_SOURCES = \
 	src/tcp.h \
 	src/text.c
 
+axel_CFLAGS = $(AM_CFLAGS) $(PTHREAD_CFLAGS)
+AM_CFLAGS = $(WARN_CFLAGS) \
+	-Wno-declaration-after-statement \
+	-Wno-error=cast-align \
+	-Wno-error=inline
+
 if WITH_SSL
 axel_SOURCES += \
 	src/dn-match.c \
 	src/ssl.c \
 	src/ssl_verify.c
+AM_CFLAGS += $(SSL_CFLAGS)
 endif
 
-AM_CFLAGS = $(PTHREAD_CFLAGS) $(WARN_CFLAGS) \
-	-Wno-declaration-after-statement \
-	-Wno-error=cast-align \
-	-Wno-error=inline
-
 axel_LDADD = $(LIBOBJS) $(SSL_LIBS) $(LIBINTL) $(PTHREAD_LIBS)
 axel_CC = $(PTHREAD_CC)
 
diff --git a/src/axel.c b/src/axel.c
index c9ca02a..88be74e 100644
--- a/src/axel.c
+++ b/src/axel.c
@@ -123,7 +123,6 @@ axel_new(conf_t *conf, int count, const search_t *res)
 {
 	axel_t *axel;
 	int status;
-	uint64_t delay;
 	url_t *u;
 	char *s;
 	int i;
@@ -151,11 +150,12 @@ axel_new(conf_t *conf, int count, const search_t *res)
 					     _("Buffer resized for this speed."));
 			axel->conf->buffer_size = axel->conf->max_speed;
 		}
-		delay = 1000000000 * axel->conf->buffer_size *
+		uint64_t delay =
+			UINT64_C(1073741824) * axel->conf->buffer_size *
 			axel->conf->num_connections / axel->conf->max_speed;
 
-		axel->delay_time.tv_sec  = delay / 1000000000;
-		axel->delay_time.tv_nsec = delay % 1000000000;
+		axel->delay_time.tv_sec  = delay / 1073741824;
+		axel->delay_time.tv_nsec = delay % 1073741824;
 	}
 	if (buffer == NULL) {
 		buffer = malloc(axel->conf->buffer_size);
diff --git a/src/axel.h b/src/axel.h
index e7323e5..6c989e5 100644
--- a/src/axel.h
+++ b/src/axel.h
@@ -148,6 +148,7 @@ void axel_close(axel_t *axel);
 void print_messages(axel_t *axel);
 
 double axel_gettime(void);
+ssize_t axel_rand64(uint64_t *);
 
 #define DN_MATCH_MALFORMED -1
 int dn_match(const char *hostname, const char *pat, size_t pat_len);
diff --git a/src/conf.c b/src/conf.c
index 75fa400..4343512 100644
--- a/src/conf.c
+++ b/src/conf.c
@@ -52,7 +52,7 @@
 	else if (!strcmp(key, #name))	\
 		dst = &conf->name;
 
-int parse_interfaces(conf_t *conf, char *s);
+static int parse_interfaces(conf_t *conf, char *s);
 
 #ifdef __GNUC__
 __attribute__((format(scanf, 2, 3)))
@@ -76,6 +76,21 @@ axel_fscanf(FILE *fp, const char *format, ...)
 	return ret;
 }
 
+static int
+parse_progress_style(conf_t *conf, const char *value)
+{
+	if (!strcasecmp(value, "auto")) {/* no-op */}
+	else if (!strcasecmp(value, "alternative"))
+		conf->progress_style = AXEL_PROGRESS_STYLE_ALTERNATIVE;
+	else if (!strcasecmp(value, "classic"))
+		conf->progress_style = AXEL_PROGRESS_STYLE_CLASSIC;
+	else if (!strcasecmp(value, "percent"))
+		conf->progress_style = AXEL_PROGRESS_STYLE_PERCENTAGE;
+	else
+		fprintf(stderr, _("Unknown progress style %s\n"), value);
+	return 1;
+}
+
 static int
 parse_protocol(conf_t *conf, const char *value)
 {
@@ -155,8 +170,6 @@ conf_loadfile(conf_t *conf, const char *file)
 			KEY(buffer_size)
 			KEY(max_speed)
 			KEY(verbose)
-			KEY(alternate_output)
-			KEY(percentage)
 			KEY(insecure)
 			KEY(no_clobber)
 			KEY(search_timeout)
@@ -188,6 +201,9 @@ conf_loadfile(conf_t *conf, const char *file)
 		else if (strcmp(key, "interfaces") == 0) {
 			if (parse_interfaces(conf, value))
 				continue;
+		} else if (strcmp(key, "progress_style") == 0) {
+			if (parse_progress_style(conf, value))
+				continue;
 		} else if (strcmp(key, "use_protocol") == 0) {
 			if (parse_protocol(conf, value))
 				continue;
@@ -266,9 +282,11 @@ conf_init(conf_t *conf)
 
 	conf->interfaces->next = conf->interfaces;
 
-    /* Detect if stdout is a tty, set the default indicator to alternate.
-       Otherwise, keep it to original.*/
-    conf->alternate_output = isatty(STDOUT_FILENO);
+	/* Detect if stdout is a tty, set the default indicator to alternate.
+	 * Otherwise, keep it to original. */
+	conf->progress_style = isatty(STDOUT_FILENO)
+		? AXEL_PROGRESS_STYLE_ALTERNATIVE
+		: AXEL_PROGRESS_STYLE_CLASSIC;
 
 	if ((s2 = getenv("http_proxy")) || (s2 = getenv("HTTP_PROXY")))
 		strlcpy(conf->http_proxy, s2, sizeof(conf->http_proxy));
@@ -306,6 +324,7 @@ conf_free(conf_t *conf)
 	free(conf->interfaces);
 }
 
+static
 int
 parse_interfaces(conf_t *conf, char *s)
 {
@@ -331,7 +350,6 @@ parse_interfaces(conf_t *conf, char *s)
 		return 1;
 	}
 
-	s[strlen(s) + 1] = 0;
 	conf->interfaces = iface = malloc(sizeof(if_t));
 	if (!conf->interfaces)
 		return 0;
diff --git a/src/conf.h b/src/conf.h
index 126dad0..74960e2 100644
--- a/src/conf.h
+++ b/src/conf.h
@@ -57,10 +57,13 @@ typedef struct {
 	int buffer_size;
 	unsigned long long max_speed;
 	int verbose;
-	int alternate_output;
 	int insecure;
 	int no_clobber;
-	int percentage;
+	enum {
+		AXEL_PROGRESS_STYLE_CLASSIC,
+		AXEL_PROGRESS_STYLE_ALTERNATIVE,
+		AXEL_PROGRESS_STYLE_PERCENTAGE,
+	} progress_style;
 
 	if_t *interfaces;
 
diff --git a/src/conn.c b/src/conn.c
index 4e6761b..c24110a 100644
--- a/src/conn.c
+++ b/src/conn.c
@@ -47,6 +47,7 @@
 
 #include "config.h"
 #include "axel.h"
+#include "hash.h"
 
 /**
  * Convert an URL to a conn_t structure.
@@ -185,7 +186,7 @@ scheme_from_proto(int proto)
 }
 
 /* Generate a nice URL string. */
-char *
+int
 conn_url(char *dst, size_t len, conn_t *conn)
 {
 	const char *prefix = "", *postfix = "";
@@ -194,7 +195,7 @@ conn_url(char *dst, size_t len, conn_t *conn)
 
 	size_t scheme_len = strlcpy(dst, scheme, len);
 	if (scheme_len > len)
-		return NULL;
+		return -1;
 
 	len -= scheme_len;
 
@@ -203,7 +204,7 @@ conn_url(char *dst, size_t len, conn_t *conn)
 	if (*conn->user != 0 && strcmp(conn->user, "anonymous") != 0) {
 		int plen = snprintf(p, len, "%s:%s@", conn->user, conn->pass);
 		if (plen < 0)
-			return NULL;
+			return -1;
 		len -= plen;
 		p += plen;
 	}
@@ -213,11 +214,8 @@ conn_url(char *dst, size_t len, conn_t *conn)
 		postfix = "]";
 	}
 
-	int plen;
-	plen = snprintf(p, len, "%s%s%s:%i%s%s", prefix, conn->host, postfix,
-			conn->port, conn->dir, conn->file);
-
-	return plen < 0 ? NULL : dst;
+	return snprintf(p, len, "%s%s%s:%i%s%s", prefix, conn->host, postfix,
+	                conn->port, conn->dir, conn->file);
 }
 
 /* Simple... */
@@ -371,6 +369,53 @@ conn_info_ftp(conn_t *conn)
 	return 1;
 }
 
+struct urlseq {
+	uint64_t secret[1];
+	char buf[MAX_STRING];
+	int num_urls;
+	uint32_t visited_urls[];
+};
+
+static
+void *
+urlseq_teardown(struct urlseq **u)
+{
+	free(*u);
+	return *u = NULL;
+}
+
+static
+struct urlseq *
+urlseq_init(unsigned int urlcount)
+{
+	struct urlseq *u;
+	u = malloc(sizeof(*u) + sizeof(u->visited_urls[0]) * urlcount);
+	if (!u || axel_rand64(u->secret) < (ssize_t)sizeof(u->secret[0]))
+		return urlseq_teardown(&u);
+	u->num_urls = 0;
+	return u;
+}
+
+static
+int
+urlseq_check_loop(struct urlseq *u, conn_t *conn)
+{
+	if (!u)
+		return 0;
+
+	int url_len = conn_url(u->buf, sizeof(u->buf), conn);
+	if (url_len <= 0)
+		return 0;
+
+	uint32_t url_hash = axel_hash32(u->buf, url_len, u->secret);
+	for (int j = 0; j < u->num_urls; j++) {
+		if (u->visited_urls[j] == url_hash)
+			return 1;
+	}
+	u->visited_urls[u->num_urls++] = url_hash;
+	return 0;
+}
+
 /* Get file size and other information */
 int
 conn_info(conn_t *conn)
@@ -383,6 +428,8 @@ conn_info(conn_t *conn)
 	char s[1005];
 	long long int i = 0;
 
+	struct urlseq *urlseq = urlseq_init(conn->conf->max_redirect);
+
 	do {
 		const char *t;
 
@@ -432,8 +479,17 @@ conn_info(conn_t *conn)
 			fprintf(stderr, _("Too many redirects.\n"));
 			return 0;
 		}
+
+		/* Check if the current URL has already been visited */
+		if (urlseq_check_loop(urlseq, conn)) {
+			fprintf(stderr, _("Redirect loop detected.\n"));
+			return 0;
+		}
 	} while (conn->http->status / 100 == 3);
 
+	/* Free the memory allocated for the redirect loop detection */
+	urlseq_teardown(&urlseq);
+
 	/* Check for non-recoverable errors */
 	if (conn->http->status != 416 && conn->http->status / 100 != 2)
 		return 0;
diff --git a/src/conn.h b/src/conn.h
index e382384..a3de150 100644
--- a/src/conn.h
+++ b/src/conn.h
@@ -110,7 +110,7 @@ typedef struct {
 } conn_t;
 
 int conn_set(conn_t *conn, const char *set_url);
-char *conn_url(char *dst, size_t len, conn_t *conn);
+int conn_url(char *dst, size_t len, conn_t *conn);
 void conn_disconnect(conn_t *conn);
 int conn_init(conn_t *conn);
 int conn_setup(conn_t *conn);
diff --git a/src/hash.c b/src/hash.c
new file mode 100644
index 0000000..095fb35
--- /dev/null
+++ b/src/hash.c
@@ -0,0 +1,146 @@
+/* Axel -- A lighter download accelerator for Linux and other Unices
+ *
+ * SipHash-2-4-32 hashing function
+ *
+ * Copyright 2016 Jean-Philippe Aumasson <jeanphilippe.aumasson@gmail.com>
+ * Copyright 2022 Ismael Luceno <ismael@iodev.co.uk>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+
+ * In addition, as a special exception, the copyright holders give
+ * permission to link the code of portions of this program with the
+ * OpenSSL library under certain conditions as described in each
+ * individual source file, and distribute linked combinations including
+ * the two.
+ *
+ * You must obey the GNU General Public License in all respects for all
+ * of the code used other than OpenSSL. If you modify file(s) with this
+ * exception, you may extend this exception to your version of the
+ * file(s), but you are not obligated to do so. If you do not wish to do
+ * so, delete this exception statement from your version. If you delete
+ * this exception statement from all source files in the program, then
+ * also delete it here.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * SPDX-License-Identifier: GPL-2.0-or-later
+ */
+
+#include <stddef.h>
+#include <stdint.h>
+#include "hash.h"
+
+/* default: SipHash-2-4 */
+#define cROUNDS 2
+#define dROUNDS 4
+
+#define FORCE_INLINE __attribute__((always_inline)) inline
+
+FORCE_INLINE static
+uint32_t
+ROTL(uint32_t x, uint8_t b)
+{
+	return x << b | x >> (32 - b);
+}
+
+static inline
+uint32_t
+U8TO32_LE(const uint8_t *p)
+{
+	return p[0] | p[1] << 8 | p[2] << 16 | p[3] << 24;
+}
+
+/**
+ * Convert from/to little endian.
+ */
+FORCE_INLINE
+uint32_t
+cu32swap(uint32_t n)
+{
+	const union { int i; char c; } u = { 1 };
+	return u.c ? n : n>>24 | (n>>8&0xff00) | (n<<8&0xff0000) | n<<24;
+}
+
+#define cpu_to_u32le cu32swap
+#define u32le_to_cpu cu32swap
+
+FORCE_INLINE static void SIPROUND(uint32_t v[4]);
+
+/**
+ * Computes a SipHash-like value.
+ */
+uint32_t
+axel_hash32(const void *src, size_t len, const uint64_t *kk)
+{
+	const uint8_t *ni = src;
+
+	const int tail = len & 3;
+	const unsigned char *end = ni + (len & ~3);
+	uint32_t b = len << 24;
+	uint32_t v[4];
+
+	v[0] = u32le_to_cpu(kk[0]);
+	v[1] = u32le_to_cpu(kk[0] >> 32);
+	v[2] = UINT32_C(0x6c796765) ^ v[0];
+	v[3] = UINT32_C(0x74656462) ^ v[1];
+
+	for (; ni != end; ni += 4) {
+		uint32_t m = U8TO32_LE(ni);
+		v[3] ^= m;
+
+		for (int i = 0; i < cROUNDS; ++i)
+			SIPROUND(v);
+
+		v[0] ^= m;
+	}
+
+	switch (tail) {
+	case 3: b |= ((uint32_t)ni[2]) << 16; /* fallthru */
+	case 2: b |= ((uint32_t)ni[1]) << 8; /* fallthru */
+	case 1: b |= ((uint32_t)ni[0]);
+	default: break;
+	}
+
+	v[3] ^= b;
+
+	for (int i = 0; i < cROUNDS; ++i)
+		SIPROUND(v);
+
+	v[0] ^= b;
+	v[2] ^= 0xff;
+
+	for (int i = 0; i < dROUNDS; ++i)
+		SIPROUND(v);
+
+	b = v[1] ^ v[3];
+	return cpu_to_u32le(b);
+}
+
+void
+SIPROUND(uint32_t v[4])
+{
+	v[0] += v[1];
+	v[1] = ROTL(v[1], 5);
+	v[1] ^= v[0];
+	v[0] = ROTL(v[0], 16);
+	v[2] += v[3];
+	v[3] = ROTL(v[3], 8);
+	v[3] ^= v[2];
+	v[0] += v[3];
+	v[3] = ROTL(v[3], 7);
+	v[3] ^= v[0];
+	v[2] += v[1];
+	v[1] = ROTL(v[1], 13);
+	v[1] ^= v[2];
+	v[2] = ROTL(v[2], 16);
+}
diff --git a/src/hash.h b/src/hash.h
new file mode 100644
index 0000000..f39bd01
--- /dev/null
+++ b/src/hash.h
@@ -0,0 +1,5 @@
+#ifndef AXEL_HASH_H
+#include <stddef.h>
+
+uint32_t axel_hash32(const void *, size_t, const uint64_t *);
+#endif
diff --git a/src/random.c b/src/random.c
new file mode 100644
index 0000000..20a7378
--- /dev/null
+++ b/src/random.c
@@ -0,0 +1,17 @@
+#include <stdatomic.h>
+#include <unistd.h>
+#include "config.h"
+#include "axel.h"
+
+ssize_t
+axel_rand64(uint64_t *out)
+{
+	static int fd = -1;
+	if (fd == -1) {
+		int tmp = open("/dev/random", O_RDONLY);
+		int expect = -1;
+		if (!atomic_compare_exchange_strong(&fd, &expect, tmp))
+			close(tmp);
+	}
+	return read(fd, out, sizeof(*out));
+}
diff --git a/src/ssl.c b/src/ssl.c
index d0459ed..dd589eb 100644
--- a/src/ssl.c
+++ b/src/ssl.c
@@ -40,7 +40,17 @@
 /* SSL interface */
 
 #include "config.h"
+
+#ifdef HAVE_WOLFSSL
+#include <wolfssl/options.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/openssl/ssl.h>
+#include <wolfssl/openssl/err.h>
+#else
+#include <openssl/ssl.h>
 #include <openssl/err.h>
+#endif
+
 #include "axel.h"
 
 static pthread_mutex_t ssl_lock;
diff --git a/src/ssl.h b/src/ssl.h
index d137ed2..8885bbe 100644
--- a/src/ssl.h
+++ b/src/ssl.h
@@ -41,7 +41,16 @@
 #if !defined(AXEL_SSL_H) && defined(HAVE_SSL)
 #define AXEL_SSL_H
 
+#include "config.h"
+
+#ifdef HAVE_WOLFSSL
+#include <wolfssl/options.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/openssl/ssl.h>
+#else
 #include <openssl/ssl.h>
+#endif
+
 
 void ssl_init(conf_t *conf);
 SSL *ssl_connect(int fd, char *hostname);
diff --git a/src/ssl_verify.c b/src/ssl_verify.c
index 0765714..266f667 100644
--- a/src/ssl_verify.c
+++ b/src/ssl_verify.c
@@ -27,8 +27,17 @@
  */
 
 #include "config.h"
+
+#ifdef HAVE_WOLFSSL
+#include <wolfssl/options.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/openssl/ssl.h>
+#include <wolfssl/openssl/x509v3.h>
+#else
 #include <openssl/ssl.h>
 #include <openssl/x509v3.h>
+#endif
+
 #include "axel.h"
 
 inline static int
diff --git a/src/tcp.h b/src/tcp.h
index 43532c7..d78cd33 100644
--- a/src/tcp.h
+++ b/src/tcp.h
@@ -44,8 +44,14 @@
 #define AXEL_TCP_H
 
 #ifdef HAVE_SSL
+#ifdef HAVE_WOLFSSL
+#include <wolfssl/options.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/openssl/ssl.h>
+#else
 #include <openssl/ssl.h>
 #endif
+#endif
 
 typedef struct {
 	int fd;
diff --git a/src/text.c b/src/text.c
index e7fa079..eb5b4b0 100644
--- a/src/text.c
+++ b/src/text.c
@@ -189,10 +189,10 @@ main(int argc, char *argv[])
 			conf->ai_family = AF_INET;
 			break;
 		case 'a':
-			conf->alternate_output = 1;
+			conf->progress_style = AXEL_PROGRESS_STYLE_ALTERNATIVE;
 			break;
 		case 'p':
-			conf->percentage = 1;
+			conf->progress_style = AXEL_PROGRESS_STYLE_PERCENTAGE;
 			break;
 		case 'k':
 			conf->insecure = 1;
@@ -237,8 +237,7 @@ main(int argc, char *argv[])
 
 	/* disable alternate outputs and verbosity when quiet is specified */
 	if (conf->verbose < 0) {
-		conf->alternate_output = 0;
-		conf->percentage = 0;
+		conf->progress_style = AXEL_PROGRESS_STYLE_CLASSIC;
 	} else if (j > -1)
 		conf->verbose = j;
 
@@ -279,7 +278,7 @@ main(int argc, char *argv[])
 		}
 	}
 
-	if (!conf->percentage)
+	if (conf->progress_style != AXEL_PROGRESS_STYLE_PERCENTAGE)
 		printf(_("Initializing download: %s\n"), s);
 
 	if (do_search) {
@@ -399,14 +398,14 @@ main(int argc, char *argv[])
 	axel_start(axel);
 	print_messages(axel);
 
-	if (conf->alternate_output || conf->percentage) {
+	if (conf->progress_style == AXEL_PROGRESS_STYLE_ALTERNATIVE
+	    || conf->progress_style == AXEL_PROGRESS_STYLE_PERCENTAGE) {
 		putchar('\n');
-	} else {
-		if (axel->bytes_done > 0) {	/* Print first dots if resuming */
-			putchar('\n');
-			print_commas(axel->bytes_done);
-			fflush(stdout);
-		}
+	} else if (axel->bytes_done > 0) {	/* Print first dots if resuming */
+		putchar('\n');
+		print_commas(axel->bytes_done);
+		fflush(stdout);
+
 	}
 	axel->start_byte = axel->bytes_done;
 
@@ -420,10 +419,10 @@ main(int argc, char *argv[])
 		prev = axel->bytes_done;
 		axel_do(axel);
 
-		if (conf->percentage) {
+		if (conf->progress_style == AXEL_PROGRESS_STYLE_PERCENTAGE) {
 			if (!axel->message && prev != axel->bytes_done)
 				printf("%u\n", calc_percentage(axel->bytes_done, axel->size));
-		} else 	if (conf->alternate_output) {
+		} else 	if (conf->progress_style == AXEL_PROGRESS_STYLE_ALTERNATIVE) {
 			if (!axel->message && prev != axel->bytes_done)
 				print_alternate_output(axel);
 		} else if (conf->verbose > -1) {
@@ -432,7 +431,7 @@ main(int argc, char *argv[])
 		}
 
 		if (axel->message) {
-			if (conf->alternate_output == 1) {
+			if (conf->progress_style == AXEL_PROGRESS_STYLE_ALTERNATIVE) {
 				/* clreol-simulation */
 				fputs("\e[2K\r", stdout);
 			} else {
@@ -440,7 +439,7 @@ main(int argc, char *argv[])
 			}
 			print_messages(axel);
 			if (!axel->ready) {
-				if (conf->alternate_output != 1)
+				if (conf->progress_style != AXEL_PROGRESS_STYLE_ALTERNATIVE)
 					print_commas(axel->bytes_done);
 				else
 					print_alternate_output(axel);
@@ -621,7 +620,7 @@ print_alternate_output(axel_t *axel)
 	if (width < 40) {
 		fprintf(stderr,
 			_("Can't setup alternate output. Deactivating.\n"));
-		axel->conf->alternate_output = 0;
+		axel->conf->progress_style = AXEL_PROGRESS_STYLE_CLASSIC;
 
 		return;
 	}
