diff --git a/src/abaddon.cpp b/src/abaddon.cpp
index 2be54ba..aee6f4f 100644
--- a/src/abaddon.cpp
+++ b/src/abaddon.cpp
@@ -1129,7 +1129,7 @@ int main(int argc, char **argv) {
     if (std::getenv("ABADDON_NO_FC") == nullptr)
         Platform::SetupFonts();
 
-    char *systemLocale = std::setlocale(LC_ALL, "");
+    char *systemLocale = std::setlocale(LC_ALL, "en_US.UTF-8");
     try {
         if (systemLocale != nullptr) {
             std::locale::global(std::locale(systemLocale));
