diff --git a/config.c b/config.c
index f120e73..4289d8a 100644
--- a/config.c
+++ b/config.c
@@ -514,6 +514,11 @@ open_config(void)
 	FILE *file;
 	char path[256];
 
+	snprintf(path, sizeof(path), "%s/velox/velox.conf", getenv("XDG_CONFIG_HOME"));
+
+	if ((file = fopen(path, "r")))
+		goto found;
+
 	snprintf(path, sizeof(path), "%s/.velox.conf", getenv("HOME"));
 
 	if ((file = fopen(path, "r")))
