diff --git a/palemoon/app/application.ini b/palemoon/app/application.ini
index c64ed9079..b57781237 100644
--- a/palemoon/app/application.ini
+++ b/palemoon/app/application.ini
@@ -20,9 +20,9 @@
 
 [App]
 # Vendor=@MOZ_APP_VENDOR@
-Vendor=Moonchild Productions
+Vendor=Local
 # Name=@MOZ_APP_BASENAME@
-Name=Pale Moon
+Name=Web
 RemotingName=@MOZ_APP_REMOTINGNAME@
 #ifdef MOZ_APP_DISPLAYNAME
 CodeName=@MOZ_APP_DISPLAYNAME@
