diff --git a/eval.go b/eval.go
index 54f2dcb..1dd90eb 100644
--- a/eval.go
+++ b/eval.go
@@ -1927,7 +1927,14 @@ func (e *callExpr) eval(app *app, args []string) {
 		app.ui.cmdAccRight = nil
 		update(app)
 	case "cmd-delete-unix-word":
-		ind := strings.LastIndex(strings.TrimRight(string(app.ui.cmdAccLeft), " "), " ") + 1
+		if len(app.ui.cmdAccLeft) == 0 {
+			return
+		}
+		locs := reWordBeg.FindAllStringSubmatchIndex(string(app.ui.cmdAccLeft), -1)
+		if locs == nil {
+			return
+		}
+		ind := locs[len(locs)-1][3]
 		app.ui.cmdYankBuf = []rune(string(app.ui.cmdAccLeft)[ind:])
 		app.ui.cmdAccLeft = []rune(string(app.ui.cmdAccLeft)[:ind])
 		update(app)
